### DISCUSSION TITLE
How to modify the sidebar dynamically on the client?

### DISCUSSION THREAD

**Discussion by chetstone** [facebook/docusaurus#4715] [Q&A]
In Docusaurus version 2.0.0-alpha.75 I'm trying to modify an autogenerated sidebar. 

I'm writing documentation for an app that runs on iOS, Android, and the web. The documentation for each platform is almost the same, but there are a few features that are present in iOS that don't exist in the other platforms and I'd like to remove them from the sidebar. I use `react-device-detect` to find out what device the user is on. (If the user is accessing the docs from the desktop I have some buttons for her to select the platform).

I have a custom sidebar items generator:

```js
          sidebarItemsGenerator: async function({
            defaultSidebarItemsGenerator,
            ...args
          }) {
            const sidebarItems = await defaultSidebarItemsGenerator(args);
            return filterSidebarItems(sidebarItems);
          }
        },
```

And my filtering function looks like this:

```js
var { currentDevice } = require('./src/components/devices');

// Filter the sidebar items
// A label with trailing '!device' items will trigger filtering.
// If there is no sidebar_label field in a doc, filtering will not be done.
// In that case there will be no 'label' field in the item.
// Device specs if present indicate 'deny' for that device

function filterSidebarItems(items) {
  // Reverse items in categories
  const result = items
    .filter(item => {
      let matches = [];
      if (item.label && item.label.includes('!')) {
        matches = item.label.split('!');
      }
      if (matches.some(match => match === currentDevice)) {
        console.log(`Removing ${item.label}`);
        return false; // suppress item
      }
      return true;
    })
    .map(item => {
      if (item.label && item.label.includes('!')) {
        item.label = item.label.split('!')[0];
        console.log(`New item label is ${item.label}`);
      }
      if (item.type === 'category') {
        return { ...item, items: filterSidebarItems(item.items) };
      }
      return item;
    });
  console.log(`Final result: ${JSON.stringify(result)}`);
  return result;
}
```

Since there is no frontmatter item that currently exists to specify which platform a doc or category belongs to, I hack the `sidebar_label` (or `label` in the case of categories), adding '!android!web' to the end of the label to tell my filter when to hide the item. I then strip the extra device indications and write the original label to the `item.label` field in the sidebar item. This doesn't seem to work. The label that appears on the site doesn't seem to come from this field, but rather I guess it comes from the actual file frontmatter, because the rendered sidebar still includes the extra device indications.

But more importantly, I'm beginning to have a sneaking feeling that this is not possible at all: maybe the sidebar is static, created at build time, and not a React component that is dynamic on the client.

If this is the case, could someone please suggest another approach to accomplishing this?

**Comment by slorber** [comment]
Hi,

The frontmatter of the docs is not part of the sidebar JSON representation structure because we want this JSON to be lightweight as it will be downloaded to display the sidebar of all the docs of this sidebar. We have very large sites with thousands of docs in the sidebar so including thousands of time complete docs metadata in the sidebars JSON is not really an option.

However, the sidebar generator receives the doc frontmatter data as input, so you can read from there.


```js
{
  sidebarItemsGenerator: async function({defaultSidebarItemsGenerator, ...args }) {
    const {docs} = args; // docs will contain the frontmatter here!
    const docsById = _.keyBy(docs,"id"); // Index the docs by id with lodash: more convenient/performant access

    const sidebarItems = await defaultSidebarItemsGenerator(args);

    return filterSidebarItems(sidebarItems,docsById);
  }
}

// You probably want to make this recursive so that subcategory items are also filtered
function filterSidebarItems(items, docsById) {
  return items.filter(item => {
    if (item.type === "doc") {
      const doc =  docsById[item.id]
      if ( doc.frontMatter.includes("ios") {
        return false;
      }
    }
    return true;
  })
}
```

Does it make sense?

For categories, you could assign `customProps` in the _category_ file so that you could easily filter based on your custom category metadata a whole category at once. Not 100% sure this is supported already, let me know if this does not work.

https://docusaurus.io/docs/next/sidebar#passing-custom-props

One thing we could add to make this more convenient for you is to add a frontmatter like `sidebar_custom_props` to docs so that you could directly filter by `item.customProps?.platforms?.includes("ios")`, would this be useful?

**Comment by chetstone** [comment]
Thanks. That’s very helpful, I’ll look into it. But what about the other
question? Does sidebarItemsGenerator only run at build time or does it
exist on the client side where it can be triggered to run when the document
is loaded or the user selects a different platform from a set of tabs? If
so, how to do that?

**Reply by slorber** [reply]
`sidebarItemsGenerator` is in config, so it runs in NodeJS, and you can't customize it using browser apis like localstorage, querystring, URL or whatever your tab system might be using.


Think about it: Docusaurus builds a static page in `/docs/myDoc`. 
This static page contains the sidebar HTML, you must choose at build time what exactly are the sidebar items that will be in the static files.
Changing the sidebar in the frontend is not ideal and can produce jank: the initial sidebar seen by the user is the static one (computed at build time), but just after the JS / React has loaded, this sidebar is updated, producing a weird visual offset, triggering ["layout shift"](https://web.dev/cls/) in the sidebar. This would lead to bad UX.

If you really need to compute a sidebar in the client-side, you could make it so by preventing the sidebar to be rendered in NodeJS (seen this done [here](https://github.com/demisto/content-docs/pull/616#issuecomment-826973273) for build perf reasons), and do your dynamic things in the browser to filter it. This may be annoying, as the sidebar would only appear once the JS has downloaded and React was hydrated (on slow network/computer, this can a be a very noticeable delay), and you'd also lose the benefits of SEO as Google would have to run the JS to see the links of your sidebar.

Really, you should avoid doing that IMHO. If ios has a different sidebar, you should rather make `/docs/ios/myDoc` + `/docs/android/myDoc` and make sure that both pages have distinct sidebars generated at build time. You can use the same source markdown folder + 2 docs plugin instances with different configs for that maybe, where each instance has a different filter in the sidebar items generator.

Or you make a single sidebar and make it clear which items belong to specific platforms. For example https://reactnative.dev/docs/accessibilityinfo

![image](https://user-images.githubusercontent.com/749374/116913024-45e29780-ac49-11eb-86c0-7388283fe44c.png)

**Comment by chetstone** [comment]
Thanks  Sébastien for all the information and the suggestion. Hopefully
this discussion can help others trying to work out a similar issue.

If ios has a different sidebar, you should rather make /docs/ios/myDoc +
> /docs/android/myDoc and make sure that both pages have distinct sidebars
> generated at build time. You can use the same source markdown folder + 2
> docs plugin instances with different configs for that maybe, where each
> instance has a different filter in the sidebar items generator.


Having different sidebars at build time seems like the way to go. But if
the source markdown folder is shared, wouldn't this run afoul of the rule
<https://docusaurus.io/docs/sidebar#sidebar-item-doc> that you should not
assign the same doc to multiple sidebars? I was thinking to use different
folders for each platform, but symlink all the shared files to the main
folder, would that work? Then I wouldn't have to overload the frontmatter
with the sidebar choice and do any filtering of the autogenerated sidebars.
I would just work with the symlinks to make the necessary changes between
the platforms and use default autogeneration.

Thanks again

Chester

**Reply by slorber** [reply]
It's not a good idea to assign a doc to 2 sidebars inside a doc plugin instance (as Docusaurus don't know which sidebar to show when browsing the doc/url), but I believe it should work fine for multiple doc instances as each doc would have a different URL and docusaurus knows which sidebar to pick for a specific URL

I have never tried with symlinks, maybe it works, not sure

**Comment by chetstone** [comment]
It seems what you're suggesting is basically Docs Multi-Instance
<https://docusaurus.io/docs/docs-multi-instance>. So why does that
documentation say, "This feature is only useful for versioned
documentations"?

**Comment by slorber** [comment]
Yes!

This is a mistake. Didn't think of your usecase 😅


Will probably edit for "mostly useful"

Le mar. 4 mai 2021 à 20:08, Chester Wood ***@***.***> a
écrit :

**Comment by chetstone** [comment]
Thanks for all the help. Thought I'd let you know how it's going... 
I had quite a bit of trouble setting up the Docs multi-instance. I've never understood routing, so I was hoping Docusaurus would hide all that stuff. But as usual with magic, if what you're doing isn't exactly what the magicians had in mind, it can be hard to work around. Anyway, as per the usual hacker method of dealing with things you don't understand (which seems to be a lot of things these days), first, find a working example that's close to what you need, then start guessing and trying this and that. 

I started with the suggested structure of putting both platforms in the docs directory: `site/docs/ios/iosdocs` and `site/docs/android/androiddocs`. I tried all sorts of config entries for `path`, `routeBasePath`, and `id` and kept getting *Page Not Found* when trying to access the android docs. Then I decided to emulate the structure of the Docusauruss site, so moved the ios docs to `site/docs/iosdocs` and the android docs to `site/android/androiddocs`. Still couldn't find anything that worked. There was no hint in the browser console or anywhere else as to what the problem might be. The debugger (http://localhost:3000/__docusaurus/debug/routes) was no help, because the routes it showed just didn't make any sense. Understandably, because as I said, I don't understand routers. 

Finally I got it to work by adding `slug: /` to the `site/android/intro.mdx` frontmatter. This is totally unintuitive to me. I understood `slug: /` to link the page to the root of the site, and why would it fix things here--it should break things. But did I mention I don't understand routers?

Anyway, once I got all that working, I tried to symlink all the ios stuff into the android directory. That seems to work. I remembered the old XWindows *lndir* command, which creates a shadow directory structure, then symlinks all the files in there. I couldn't find an [implementation of lndir](https://github.com/enigmata/lndir) that would build on my mac, so hacked up a script to do that. After symlinking, I go in and remove the things I don't want in the android sidebar, and replace symlinks with files that have to be different (like intro.mdx which has to have the slug). This works great but I think it will be a bit messy to maintain. For my small project documenting a mature product, it's probably OK, but most projects would do well to share the markdown directory and filter the autogenerated sidebar as sketched out above. I may switch to this later but for now want to get started writing the documentation.

FYI, here's what is working now:

*docusaurus.config.js*
```js 
  presets: [
    [
      '@docusaurus/preset-classic',
      {
        docs: {
          path: 'docs',
          sidebarPath: require.resolve('./sidebars_ios.js')
         // no routeBasePath or id
        },
    },
  ],
  plugins: [
    [
      '@docusaurus/plugin-content-docs',
      {
        id: 'android',
        path: 'android',
        routeBasePath: 'android',
        sidebarPath: require.resolve('./sidebars_android.js')
        // ... other options
      }
    ]
  ]

```

*sidebars_ios.js*
```js
module.exports = {
  // By default, Docusaurus generates a sidebar from the docs folder structure
  iosSidebar: [{ type: 'autogenerated', dirName: '.' }]
}
```

*sidebars_android.js*
```js
module.exports = {
  // By default, Docusaurus generates a sidebar from the docs folder structure
  androidSidebar: [{ type: 'autogenerated', dirName: '.' }]
}
```

*site/docs/intro.mdx*
```yml
---
id: iosintro
sidebar_position: 1
title: 'Introduction'
sidebar_label: 'Introduction'
hide_title: true
---
```

*site/android/intro.mdx*
```yml
---
id: androidintro
slug: /
sidebar_position: 1
title: 'Introduction'
sidebar_label: 'Introduction'
hide_title: true
---
```

So thanks again Sébastien for all the help!

**Reply by slorber** [reply]
great that you find a solution.

I don't think you need `slug: '/'`. Your intro doc should be accessible at `/android/intro` by default (not `/android/`)

Note that we have a debug plugin in case you want to debug the routes created.

Example: https://docusaurus.io/__docusaurus/debug/routes

**Comment by chetstone** [comment]
Thanks. You're right. I removed the `slug: / ` and am able to access at `/android/androidintro`
